Add unindexed matcher capability to query

alfinkel opened the pull request:
This pull request provides simple, single level AND-able post hoc matcher functionality to complement the current set of Cloudant query SQL engine matching functionality.

@mikerhodes - points of interest in this PR are:

- The "match" implementation in QueryResult 
https://github.com/cloudant/sync-android/commit/f59f7bf3f8432fd0a1c51ced7286b03f6b8ce0f0#diff-5f5fe154c6c97af9fc67a7b9f8ae9c87R39

- The query rejection in SQLQueryTranslator when floats are found in a query object.  
https://github.com/cloudant/sync-android/commit/f59f7bf3f8432fd0a1c51ced7286b03f6b8ce0f0#diff-10b697b5119c2ac1b73822d862ec16dcR44
https://github.com/cloudant/sync-android/commit/f59f7bf3f8432fd0a1c51ced7286b03f6b8ce0f0#diff-10b697b5119c2ac1b73822d862ec16dcR172

The rest of the code is pretty much a straight port from iOS.

alfinkel commented:
@mikerhodes when you get a chance can you take a look at this PR?  Thanks.

mikerhodes commented on sync-core/src/main/java/com/cloudant/sync/query/QueryResult.java:
```diff
@@ -44,23 +42,19 @@ public QueryResult(List<String> docIds,
                        UnindexedMatcher matcher) {
-        this.docIds = docIds;
+        this.docIds = matcher == null ? processDocIds(docIds, skip, limit) :
+                                        processDocIds(docIds, datastore, skip, limit, matcher);
```
This processing needs to happen during the iteration process to avoid false positives/negatives as the contents of the document store could change between this point and the iterator.

alfinkel replied:
I re-wrote the QueryResult processing so that it more closely resembles what is being done in the iOS version.  I also added skip and limit testing since skip and limit are now implemented.  Done in 3f1b4c2.  

https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-5f5fe154c6c97af9fc67a7b9f8ae9c87L31
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-1a5092ac6698384ba572368c278436fdR311

mikerhodes commented on sync-core/src/main/java/com/cloudant/sync/query/UnindexedMatcher.java:
```diff
@@ -98,14 +106,221 @@ public static UnindexedMatcher matcherWithSelector(Map<String, Object> selector)
+            //   "not (there's an item that matches white_cat)"
+            // The latter is satisfied using the $nin operator.
+            if (actual instanceof List) {
+                boolean currentItemPassed = false;
```
Not used outside `for`, declare inside, avoid the compiler warning.

alfinkel replied:
Done in 3f1b4c2.

https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-71032f1442a4b45142f92307d7418a22L236

mikerhodes commented on sync-core/src/main/java/com/cloudant/sync/query/UnindexedMatcher.java:
```diff
@@ -98,14 +106,221 @@ public static UnindexedMatcher matcherWithSelector(Map<String, Object> selector)
+        }
+        return l instanceof Number && r instanceof Number &&
+                !(l instanceof Float || r instanceof Float) &&
+                ((Number) l).longValue() == ((Number) r).longValue();
```
Seems like this will compare `1.0` and `1.1` as equal. I think this needs making protected and some tests written.

alfinkel replied:
You are absolutely right.  I changed Number comparison from .longValue() to .doubleValue() and added tests to verify conditions.  Done in 3f1b4c2.
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-71032f1442a4b45142f92307d7418a22L284
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-9845e00b021fbb4b1f9cd0cb43e43b23R59

mikerhodes commented on sync-core/src/test/java/com/cloudant/sync/query/AbstractQueryExecutorTestBase.java:
```diff
@@ -0,0 +1,280 @@
+public abstract class AbstractQueryExecutorTestBase extends AbstractIndexTestBase {
+
+    @Override
+    public void setUp() throws SQLException {
```
I don't think this works correctly, because you appear to create the same class of IndexManager object (in `AbstractIndexTestBase` `setUp()`) for all of the query test classes, which will use both SQL indexes and the unindexed matcher?

You need sub-classes of IndexManager specific to testing, which override:

* the `QueryTranslator#translateQuery()` to always return a query which matches all document (so only the unindexed matcher is being tested).
* the `QueryTranslator#matcherForIndexCoverage()` to always return `null` (so only the SQL query engine is tested).

alfinkel replied:
I changed the processing of these tests to now use mock IndexManager and QueryExecutor subclasses to force the tests to go in two directions.  Same set of tests are executed once against the SQL engine and once against the post hoc matcher.  Done in 3f1b4c2.
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-1a5092ac6698384ba572368c278436fdL47
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-0eaf4604dd4e002edfe0d41fdea2c7e3R20
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-162b0c09f735c14ea05abac3dd234d9eR27
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-c4ac637a25d1bf100c90b6b658737b56R20
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-253a3500c1d06e06260c5b09832b8986R21
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-c8d8ef5f4948719512f62057f5359599L29
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-162fe9eab99c2b46b7150404a5da366dL28

alfinkel commented:
@mikerhodes please take a look when you have a chance at 3f1b4c2.  I believe that I addressed all of your comments for https://github.com/cloudant/sync-android/pull/81.

mikerhodes commented on sync-core/src/main/java/com/cloudant/sync/query/OperatorExpressionNode.java:
```diff
@@ -0,0 +1,32 @@
+//  Copyright (c) 2014 Cloudant. All rights reserved.
```
It's now 2015 :)

alfinkel replied:
Updated all copyright notices on classes written in 2015.  Done in https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3

mikerhodes commented on sync-core/src/test/java/com/cloudant/sync/query/QueryExecutorMatcherTest.java:
```diff
@@ -0,0 +1,39 @@
+import java.sql.SQLException;
+import java.util.Arrays;
+
+public class QueryExecutorMatcherTest extends AbstractQueryTestBase {
```
I think it's worth adding a comment to each of the four classes `MockSQLOnlyQueryExecutor`, `MockSQLOnlyIndexManager`, `MockMatcherQueryExecutor` and `MockMatcherIndexManager` to explain what each one does -- or you could point to explanations you add to `QueryExecutorMatcherTest` and `QueryExecutorSQLOnlyTest`.

This set up is rather odd -- perhaps we can figure out a way to clean up both the objc and this one sometime! -- and so needs documenting.

alfinkel replied:
Added comments where appropriate explaining what each class does.  I took a note to revisit the process later on in both iOS and Android once the smoke clears.
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-0eaf4604dd4e002edfe0d41fdea2c7e3R20
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-162b0c09f735c14ea05abac3dd234d9eR27
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-c4ac637a25d1bf100c90b6b658737b56R20
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-253a3500c1d06e06260c5b09832b8986R21
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-c8d8ef5f4948719512f62057f5359599R24
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-162fe9eab99c2b46b7150404a5da366dR24

mikerhodes commented on sync-core/src/test/java/com/cloudant/sync/query/UnindexedMatcherTest.java:
```diff
@@ -0,0 +1,394 @@
+        assertThat(UnindexedMatcher.compareEq(1, 1l), is(true));
+        assertThat(UnindexedMatcher.compareEq(1l, 2), is(false));
+        assertThat(UnindexedMatcher.compareEq(2, 1l), is(false));
+    }
```
Do you think you need to do the boxing manually instead of relying on autoboxing for the compareEq tests? Given that `compareEq` will never autobox because all its inputs will come from dictionaries where the boxing has already happened? I'd say you want to be passing Integers and Doubles in, and ensuring they work, and passing in Floats and ensuring they don't.

I'd also statically import compareEq if that's possible.

alfinkel replied:
Added UnindexMatcher.compareEq as a static import to the test class and added manual boxing to all Numbers being used during the compare equality tests.

https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-9845e00b021fbb4b1f9cd0cb43e43b23R15
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-9845e00b021fbb4b1f9cd0cb43e43b23R61

mikerhodes commented on sync-core/src/main/java/com/cloudant/sync/query/QueryResult.java:
```diff
@@ -55,109 +54,132 @@ public QueryResult(List<String> docIds,
+
+                    // Apply projection if result matches
+                    if (fields != null && !fields.isEmpty()) {
+                        // TODO - projection
```
For things that are not yet done, I suggest throwing NotImplementedException. Same for the operators and so on.

mikerhodes replied:
Though perhaps not here, now I recall where we are in the code. This TODO needs more context really. Or just remove this code until projection is implemented, I don't think it's needed even as a placeholder?

alfinkel replied:
I believe that the resolution we came to today was to leave this PR as it is wrt the handling of unimplemented functionality and that in the next set of commits and subsequent new PR to account for unimplemented functionality more gracefully there.  With that said, I've removed the if placeholder logic for projection and replaced it with a single TODO comment.
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-5f5fe154c6c97af9fc67a7b9f8ae9c87L143

mikerhodes commented on sync-core/src/main/java/com/cloudant/sync/query/QueryResult.java:
```diff
@@ -55,109 +54,132 @@ public QueryResult(List<String> docIds,
     }
 
     @Override
     public Iterator<BasicDocumentRevision> iterator() {
```
Public API should use `DocumentRevision` interface not the specific class.

alfinkel replied:
Made QueryResult implement Iterable\<DocumentRevision> and used DocumentRevision throughout the code.  The exception being the case where we are documents are retrieved from the datastore since the datastore returns List\<BasicDocumentRevision>.
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-5f5fe154c6c97af9fc67a7b9f8ae9c87L29

mikerhodes commented on sync-core/src/main/java/com/cloudant/sync/query/OperatorExpressionNode.java:
```diff
@@ -0,0 +1,32 @@
+ *
+ *  @see com.cloudant.sync.query.UnindexedMatcher
+ */
+public class OperatorExpressionNode implements QueryNode {
```
Private

alfinkel replied:
I made the class package protected instead of public.  In order to make it private it would have to be an inner class.
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-cac498342a06ebddae3efd7af8bbbf5eL28
I did consider moving all of the QueryNode classes to be inner classes of the QueryExecutor.  @mikerhodes what are your thoughts on that?  Would it be better to leave them as top level classes or does it make sense to make them inner to the executor?  I go back and forth with what I think is right.

mikerhodes commented on sync-core/src/test/java/com/cloudant/sync/query/AbstractQueryTestBase.java:
```diff
@@ -105,13 +132,38 @@ public void setUp() throws SQLException {
+        Map<String, Object> query = new HashMap<String, Object>();
+        query.put("name", nameOperator);
+        Map<String, Object> ageOperator = new HashMap<String, Object>();
+        ageOperator.put("$eq", 12.0f);
```
If `12.0f` is the invalid value (because it's a float) note that somewhere in the test.

alfinkel replied:
Added a comment to the test description.
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-1a5092ac6698384ba572368c278436fdR157

mikerhodes commented on sync-core/src/test/java/com/cloudant/sync/query/AbstractQueryTestBase.java:
```diff
@@ -17,27 +17,55 @@
+ *  {@link com.cloudant.sync.query.QueryExecutorMatcherTest} which in turn
+ *  exercises the post hoc matcher matching functionality.
+ */
+public abstract class AbstractQueryTestBase {
```
You still need a QueryExecutorTest class which tests the default implementations of IndexManager and QueryExecutor. Right now you only have the two classes which test each separately, you need the test for the use of both. The `QueryExecutorTest` might need some extra tests to cover the cases where a combination of both SQL and unindexed matcher is needed -- because the Abstract class probably only has tests which can be solved by either in isolation (depending on the indexes defined).

alfinkel replied:
Added a `QueryExecutorStandardTest` which uses the regular `IndexManager` and `QueryExecutor`. 
https://github.com/cloudant/sync-android/commit/cf527fd540554a34338857f179c44fa358444ad3#diff-ce1fd18195cc9f32262108611d0dd922R25
As of right now though only simple AND query is supported so only either the matcher or the SQL engine is being used to satisfy the results of a query.  In the next set of commits and PR I will be providing support for OR and subclauses so there will be a need at that point to add more tests so that the "standard" way of querying is validated appropriately through `QueryExecutorStandardTest`.

mikerhodes commented on sync-core/src/main/java/com/cloudant/sync/query/UnindexedMatcher.java:
```diff
@@ -98,14 +106,221 @@ public static UnindexedMatcher matcherWithSelector(Map<String, Object> selector)
+        // TODO - Handle OR subclauses
+
+        // Add subclauses that are AND
+        // TODO - Handle AND subclauses
```
I think the query should fail if the user passed in something with `and/or` subclauses -- or do you check for this in validation?

alfinkel replied:
Copying response from above:
I believe that the resolution we came to today was to leave this PR as it is wrt the handling of unimplemented functionality and that in the next set of commits and subsequent new PR to account for unimplemented functionality more gracefully there.

mikerhodes commented on sync-core/src/main/java/com/cloudant/sync/query/QueryResult.java:
```diff
@@ -55,109 +55,129 @@ public QueryResult(List<String> docIds,
+
+                if (stop) {
+                    // Limit reached.  Move range location to the end.
+                    range.location = originalDocIds.size();
```
I'd not use this technique of pretending the end of the list has been reached when really you've reached the limit, instead I'd make `stop` class level and make the hasNext check include it: 

```java
if (stop) {
    return false;
} else {
    return (documentBlock.hasNext() || range.location < originalDocIds.size());
}
```

Perhaps renamed `stop` to `limitReached`.

alfinkel replied:
Done.https://github.com/cloudant/sync-android/commit/beba6a1c5a2456f5dfdf0ff8f508d931d52ba34c#diff-5f5fe154c6c97af9fc67a7b9f8ae9c87R86
